compiler/rustc_resolve/src/macros.rs

```diff
@@ -140,10 +142,13 @@ pub(crate) fn registered_tools(tcx: TyCtxt<'_>, (): ()) -> RegisteredTools {
             }
         }
     }
-    // We implicitly add `rustfmt` and `clippy` to known tools,
+    // We implicitly add `rustfmt`, `clippy`, `diagnostic` to known tools,
     // but it's not an error to register them explicitly.
     let predefined_tools = [sym::clippy, sym::rustfmt];
     registered_tools.extend(predefined_tools.iter().cloned().map(Ident::with_dummy_span));
+    if tcx.features().diagnostic_namespace {
+        registered_tools.insert(Ident::with_dummy_span(sym::diagnostic));
+    }
```

Comment:
This is ok to land on this PR as is, but I believe that as we iterate on the implementation we'll have to move the restriction of the `diagnostic` namespace away from here and onto a later stage, so that the error is not about "diagnostic not found" but rather "you're using diagnostic which is nightly only".